### ISSUE TITLE
src build on windows not working

### ISSUE THREAD

**Issue by gregorsoll** [protocolbuffers/protobuf#12947]
**What version of protobuf and what language are you using?**
Version: main/v3.6.0/v3.5.0 etc. (NOTE: please try updating to the latest version of protoc/runtime possible beforehand to attempt to resolve your problem)
Language: C++/Java/Python/C#/Ruby/PHP/Objective-C/Javascript

**What operating system (Linux, Windows, ...) and version?**
Windows 

**What runtime / compiler are you using (e.g., python version or gcc version)**
VC 14.16.27023

**What did you do?**
1. cloned src and submodules
2. installed bazel, c++ buildtools 
3. bazel build :protoc :protobuf

**What did you expect to see**
build completed

**What did you see instead?**
 C:/users/xyz/_bazel_xyz/y7fcwcs2/external/com_google_protobuf/src/google/protobuf/compiler/csharp/BUILD.bazel:22:11: Compiling src/google/protobuf/compiler/csharp/csharp_repeated_enum_field.cc [for tool] failed: (Exit 2): cl.exe failed: error executing command (from target @com_google_protobuf//src/google/protobuf/compiler/csharp:csharp) C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\cl.exe /nologo /DCOMPILER_MSVC /DNOMINMAX /D_WIN32_WINNT=0x0601 /D_CRT_SECURE_NO_DEPRECATE ... (remaining 65 arguments skipped)
cl : Command line warning D9002 : ignoring unknown option '-std=c++14'
external/com_google_protobuf/src/google/protobuf/compiler/csharp/csharp_repeated_enum_field.cc(31): fatal error C1083: Cannot open include file: 'google/protobuf/compiler/csharp/csharp_repeated_enum_field.h': No such file or directory
ERROR: C:/users/xyz/_bazel_xyz/y7fcwcs2/external/com_google_protobuf/src/google/protobuf/compiler/BUILD.bazel:74:11: Compiling src/google/protobuf/compiler/command_line_interface.cc [for tool] failed: (Exit 2): cl.exe failed: error executing command (from target @com_google_protobuf//src/google/protobuf/compiler:command_line_interface) C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\cl.exe /nologo /DCOMPILER_MSVC /DNOMINMAX /D_WIN32_WINNT=0x0601 /D_CRT_SECURE_NO_DEPRECATE ... (remaining 68 arguments skipped)
cl : Command line warning D9002 : ignoring unknown option '-std=c++14'
external/com_google_protobuf/src/google/protobuf/compiler/command_line_interface.cc(35): fatal error C1083: Cannot open include file: 'google/protobuf/compiler/command_line_interface.h': No such file or directory
ERROR: C:/users/xyz/_bazel_xyz/y7fcwcs2/external/com_google_protobuf/src/google/protobuf/compiler/objectivec/BUILD.bazel:53:11: Compiling src/google/protobuf/compiler/objectivec/primitive_field.cc [for tool] failed: (Exit 2): cl.exe failed: error executing command (from target @com_google_protobuf//src/google/protobuf/compiler/objectivec:objectivec) C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\cl.exe /nologo /DCOMPILER_MSVC /DNOMINMAX /D_WIN32_WINNT=0x0601 /D_CRT_SECURE_NO_DEPRECATE ... (remaining 67 arguments skipped)
cl : Command line warning D9002 : ignoring unknown option '-std=c++14'
external/com_google_protobuf/src/google/protobuf/compiler/objectivec/primitive_field.cc(31): fatal error C1083: Cannot open include file: 'google/protobuf/compiler/objectivec/primitive_field.h': No such file or directory
INFO: Elapsed time: 4.749s, Critical Path: 0.59s
INFO: 29 processes: 29 internal.
FAILED: Build did NOT complete successfully

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

**Anything else we should know about your project / environment**

**Comment by deannagarcia** [comment]
It looks like the protobuf version you're using is pretty old and is no longer supported. Can you upgrade to the newest version (23.2) and see if you still see the issue?

**Comment by laramiel** [comment]
I ran into the same issue trying to upgrade to 23.x and here's my diagnosis:

MSVC cl.exe has a limitation (https://developercommunity.visualstudio.com/t/clexe-compiler-driver-cannot-handle-long-file-path/975889) on filename lengths of ~256 characters (I think, it may be a bit smaller due to drive, etc).

Within Protobuf there are a number of source files organized like this:

https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/compiler/objectivec/BUILD.bazel

with rules like (abbreviated):

cc_library(
    name = "objectivec",
    srcs = [ ... ],
    hdrs = [
        ...
        "text_format_decode_data.h",
    ],
    include_prefix = "google/protobuf/compiler/objectivec",
)

When bazel encounters rules like this it constructs output paths similar to this:

`<bazel_output_base>/<hash>/execroot/<root workspace>/bazel-out/x64_windows-<config>/bin/external/com_google_protobuf/src/google/protobuf/compiler/objectivec/_virtual_includes/objectivec/google/protobuf/compiler/objectivec/text_format_decode_data.h`

Counting the hash, but using 'fastbuild' for the config, but not counting <bazel_outout_base> & <root workspace>, this filename is already at 215 characters. 

That leaves a maximum of 40 characters for the other parts.

It's pretty easy to run out of that. The default output base for my username is >30 characters.

One option to fix this is to reorganize the src/google/protobuf/compiler files to remove the nesting.

**Comment by laramiel** [comment]
https://github.com/bazelbuild/bazel/issues/15961

**Comment by laramiel** [comment]
I tried to rename my dependency from `com_google_protobuf` to `protobuf`. There are two things which encumber that.

1. The built-in repository `@bazel_tools` references `@com_google_protobuf://protoc`. There does not appear to be a way to override the `repo_mapping` for this repository to make it reference something else.

There is a repository https://github.com/bazelbuild/rules_proto which maybe bazel is using to externalize proto rules, but it doesn't seem very active. I'm not sure of the state of that.

2. grpc binds 'protobuf' already. This may not be the place to lament the style of the grpc bazel rules, but their extensive use of bind complicates my attempt to get a hermetic environment where my repository controls all the dependencies and does not allow dependencies to load third-party repositories.

So it turns out that it's a bit challenging to save characters simply by renaming the repository.

**Comment by hlopko** [comment]
@meteorcloudy could you recommend the best solution for us?

**Comment by meteorcloudy** [comment]
Using a shorter output base / output user root is the only direct workaround I can think of as recommended at https://bazel.build/configure/windows#long-path-issues

@oquenchil Is it possible to create the `virtual_includes` dir under some path that the length is `O(1)` instead of `O(package path length)`?

**Comment by jskeet** [comment]
@hlopko: I'm afraid I'm really not the person to ask about this - my understanding of both C++ and Bazel is pretty low. I have no objection to it being changed, but it's not something I'd have confidence either doing or reviewing.

**Comment by meteorcloudy** [comment]
I guess we can at least try and see if it breaks any tests? ;)

**Comment by oquenchil** [comment]
> @oquenchil Is it possible to create the virtual_includes dir under some path that the length is O(1) instead of O(package path length)?

I don't think so. 
- Assuming source code cannot be changed the directory structure must be as listed on the include directive
- The symlink must be in a folder belonging to the package of the target creating the symlink

That doesn't leave out anything that can be shortened using a hash like we do for shared libraries on Windows.

I can't think of anything except what's already been mentioned: not using `include_prefix` or shorter output base.

**Comment by laramiel** [comment]
I think that this would require a bazel change; basically bazel would need to be able to hand out such a directory with a shorter prefix. I don't know what that would entail.

I filed a feature request for that: https://github.com/bazelbuild/bazel/issues/18683

It was recategorized to c++ rules, which may not be a great category.

**Comment by hlopko** [comment]
Looking around, it seems that the reason to have all the 73 instances if include_prefix in the repo is easier compatibility with the Google internal clone of this repository. One solution here would be to fix header include paths in all C++ sources so we don't need include_prefix at all.

CC @mkruskal-google

**Comment by zhangyanjun2020** [comment]
邮件已收到

**Comment by zhangyanjun2020** [comment]
邮件已收到

**Comment by zhangyanjun2020** [comment]
邮件已收到

**Comment by zhangyanjun2020** [comment]
邮件已收到

**Comment by freeform-andre** [comment]
I'm running into this issue on `v28.3` even with a short output base. The only other cause I can think of is that I'm building without runfiles enabled as my windows environment does not have symlinks enabled. If there is a way to eliminate the virtual imports or fix whatever issue in upstream `rules_cc` that'd be fantastic.

**Comment by zhangyanjun2020** [comment]
邮件已收到

**Comment by zhangyanjun2020** [comment]
邮件已收到

**Comment by neilconway** [comment]
@esrauchg The `engdoc.corp.google.com` link you posted is not publicly accessible.

**Comment by esrauchg** [comment]
Apologies, accidentally used the incorrect url: corrected to https://protobuf.dev/news/v30/#poison-msvc--bazel

**Comment by crt-31** [comment]
> and because bazel+msvc usage is low compared to other ways to build on Windows, 

Um, I don't know if this is necessarily a proper statement.  Regardless, bazel+msvc should be a supported configuration as it has been. Please don't 'drop support'... lets find a way forward, I know we can.

**Comment by Mizux** [comment]
Should be fixed now see:
* https://github.com/protocolbuffers/protobuf/issues/20085

TLDR: Need Bazel=8.4.0+ and rules_cc=0.1.4+

**Comment by zhangyanjun2020** [comment]
邮件已收到

**Comment by esrauchg** [comment]
As clarified on https://github.com/protocolbuffers/protobuf/issues/20085, Bazel managed to free up a lot more characters which fixed the underlying infra problem that was the root cause of the breakage here, and correspondingly its now possible for Protobuf to sensibly compile in that environment.

If you still see path length related issues on latest protobuf and latest bazel, please open a new issue.

**Comment by zhangyanjun2020** [comment]
邮件已收到
